### DISCUSSION TITLE
Should we limit cross-thread segment count, just like the span limit?

### DISCUSSION THREAD

**Discussion by shichaoyuan** [apache/skywalking#11408] [General]
The number of spans in the segment is limited in `TracingContext`, which does not exceed 800 by default.

![image](https://github.com/apache/skywalking/assets/1159482/84c14c0a-0ca7-4d8b-baf7-d86907eb5c8f)

This mechanism does not limit the number of cross-thread.

Should we limit cross-thread segment count, just like the span limit?

**Comment by wu-sheng** [comment]
Generally, we don't want to, because it could be a one-to-many batch task.
Why do you raise this? Any specific case?

**Reply by shichaoyuan** [reply]
In some scenarios where callbacks trigger scheduled tasks, thousands of cross-thread tasks will be created, which will generate a large number of segments.
But in fact the information is of little value, we want a mechanism to limit the number of cross-thread segments collected.

**Reply by wu-sheng** [reply]
The worth case to limit this would cost more resources.
How do you think to limit the number of tracing context without damaging the agent performance in other cases?

**Reply by shichaoyuan** [reply]
Is it possible to limit the number of capture `ContextSnapshot`, and capture `IgnoredContextSnapshots` if it is exceeded?
![image](https://github.com/apache/skywalking/assets/1159482/bcc1ca04-858c-43ae-8001-cd9caa55bc96)

**Reply by wu-sheng** [reply]
How are you possible to know the threshold to set?

**Reply by shichaoyuan** [reply]
This threshold may depend on the respective needs and scenarios, just like `spanLimit` in `SpanLimitWatcher`, which should also be an empirical value.

**Reply by wu-sheng** [reply]
I mean, from plugin dev perspective, how to determine this value?
For example, about which plugin, you are confident to set the value?

**Reply by shichaoyuan** [reply]
Plugin developers do not need to pay attention to this value, only end users may need to pay attention.

**Reply by wu-sheng** [reply]
How to pay attention.
Exposing configurations is not the way to resolve issues, the solution of how to set this value is.
